agent/hcp/telemetry_config_provider.go

```diff
@@ -0,0 +1,203 @@
+)
+
+const (
+	// defaultRefreshInterval is a default duration, after each
+	defaultRefreshInterval = 5 * time.Minute
+)
+
+var (
+	// internalMetricRefreshFailure is a metric to monitors refresh failures.
+	internalMetricRefreshFailure []string = []string{"hcp", "telemetry_config_provider", "refresh", "failure"}
+)
+
+// TelemetryConfigProviderOpts is used to initialize a telemetryConfigProvider.
+type TelemetryConfigProviderOpts struct {
+	ctx             context.Context
+	endpoint        *url.URL
+	labels          map[string]string
+	filters         *regexp.Regexp
+	refreshInterval time.Duration
+	cloudCfg        config.CloudConfig
+	hcpClient       hcpclient.Client
+}
+
+// metricsConfig is a set of configurable settings for metrics collection, processing and export.
+type metricsConfig struct {
+	endpoint *url.URL
+	labels   map[string]string
+	filters  *regexp.Regexp
+}
+
+// telemetryConfigProvider holds metrics configuration and settings for its continuous
+// fetch of new config from HCP.
+type telemetryConfigProvider struct {
+	// metricsConfig holds metrics configuration that can be dynamically updated
+	// based on updates fetched from HCP.
+	metricsConfig *metricsConfig
+
+	// refreshInterval controls the interval at which new configuration is fetched from HCP.
+	refreshInterval time.Duration
+
+	// a reader-writer mutex is used as the provider is read heavy, as the OTEL components
+	// access metricsConfig, while config is only updated (write) when there are changes.
+	rw sync.RWMutex
+
+	logger    hclog.Logger
+	cloudCfg  config.CloudConfig
+	hcpClient hcpclient.Client
+}
+
+func NewTelemetryConfigProvider(opts *TelemetryConfigProviderOpts) *telemetryConfigProvider {
+	m := &metricsConfig{
+		endpoint: opts.endpoint,
+		labels:   opts.labels,
+		filters:  opts.filters,
+	}
+
+	t := &telemetryConfigProvider{
+		metricsConfig:   m,
+		logger:          hclog.FromContext(opts.ctx).Named("telemetry_config_provider"),
+		refreshInterval: opts.refreshInterval,
+		cloudCfg:        opts.cloudCfg,
+		hcpClient:       opts.hcpClient,
+	}
+
+	go t.run(opts.ctx)
+
+	return t
+}
+
+// run continously checks for updates to the telemetry configuration by
+// making a request to HCP, and verifying for any updated fields.
+func (t *telemetryConfigProvider) run(ctx context.Context) {
+	ticker := time.NewTicker(t.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if m, hasChanged := t.checkUpdate(ctx); hasChanged {
+				// Only update metricsConfig changes are detected
+				// to decrease usage of write locks that block read locks.
+				t.modifyMetricsConfig(m)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *telemetryConfigProvider) checkUpdate(ctx context.Context) (*metricsConfig, bool) {
```

Comment:
Maybe it would be simpler to do this on the TWG side and return a "hasChanged" field, so that Consul knows there have been changes more easily?